hail/python/hailtop/batch/job.py

```diff
@@ -583,3 +374,542 @@ def _pretty(self):
+                                     f"Hint: resources must be from the same batch as the current job.")
+
+            if r._source != self:
+                self._add_inputs(r)
+                if r._source is not None:
+                    if r not in r._source._valid:
+                        name = r._source._resources_inverse[r]
+                        raise BatchException(f"undefined resource '{name}'\n"
+                                             f"Hint: resources must be defined within "
+                                             f"the job methods 'command' or 'declare_resource_group'")
+                    self._dependencies.add(r._source)
+                    r._source._add_internal_outputs(r)
+            else:
+                _add_resource_to_set(self._valid, r)
+            self._mentioned.add(r)
+            return f"${{{r_uid}}}"
+
+        regexes = [_resource.ResourceFile._regex_pattern,
+                   _resource.ResourceGroup._regex_pattern,
+                   _resource.PythonResult._regex_pattern,
+                   Job._regex_pattern,
+                   batch.Batch._regex_pattern]
+        subst_command = re.sub('(' + ')|('.join(regexes) + ')',
+                               handler,
+                               command)
+        self._command.append(subst_command)
+        return self
+
+
+class PythonJob(Job):
+    """
+    Object representing a single job to execute.
+
+    Examples
+    --------
+
+    Create a new Python job that multiplies two numbers and then adds 5 to the result:
+
+    .. code-block:: python
+
+        # Create a batch object with a default Python image
+
+        b = Batch(default_python_image='gcr.io/hail-vdc/python-dill:3.7-slim')
+
+        def multiply(x, y):
+            return x * y
+
+        def add(x, y):
+            return x + y
+
+        j = b.new_python_job()
+        j.call(multiply, 2, 3)
+        j.call(add, result, 5)
+
+        # Write out the str representation of result to a file
+
+        b.write_output(result.as_str(), 'hello.txt')
+
+        b.run()
+
+    Notes
+    -----
+    This class should never be created directly by the user. Use :meth:`.Batch.new_python_job`
+    instead.
+
+    The image to use for Python jobs can be specified by `default_python_image`
+    when constructing a :class:`.Batch`. The image specified must have the `dill`
+    package installed. If ``default_python_image`` is not specified, then a Docker
+    image will automatically be created for you with the base image
+    `hailgenetics/python-dill:[major_version].[minor_version]-slim` and the Python
+    packages specified by ``python_requirements`` will be installed. The default name
+    of the image is `batch-python` with a random string for the tag unless ``python_build_image_name``
+    is specified. If the :class:`.ServiceBackend` is the backend, the locally built
+    image will be pushed to the repository specified by ``image_repository``.
```

Comment:
These aren't fields in `PythonJob`, so this explanation might be better placed on `Batch` or somewhere adjacent.

Reply:
Moved to `Batch.new_python_job`